### ISSUE TITLE
cudaPackages.nsight_systems: `nsys-ui` binary fails to start

### ISSUE THREAD

**Issue by mcwitt** [NixOS/nixpkgs#232458]
### Describe the bug
The Nsight Systems UI binary fails to start with the following error:

```
Error: Nsight Systems #VERSION_RSPLIT# hasn't been installed with CUDA Toolkit #CUDA_MAJOR#.#CUDA_MINOR#
```

### Steps To Reproduce
``` shell
nix shell --impure --expr 'with builtins; with import (getFlake github:NixOS/nixpkgs/master) { config = { allowUnfree = true; cudaSupport = true; }; }; cudaPackages.nsight_systems' --command 'nsys-ui'
```

### Notify maintainers

@ConnorBaker @samuela @SomeoneSerge 

### Metadata
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.27, NixOS, 23.05 (Stoat), 23.05.20230506.897876e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by SomeoneSerge** [comment]
I didn't check the gui, but running this on a remote machine doesn't result in any crashes for me:

```console
❯ nix shell --impure --expr 'with import (builtins.getFlake github:NixOS/nixpkgs/master) { config.allowUnfree = true; config.cudaSupport = true; }; cudaPackages.cudatoolkit' --command 'nsys-ui'
OpenGL version: "3.1 Mesa 18.1.9 (git-f57f37f3ba)"
```

Side note: it's preferable to use redist `cudaPackages` instead of `cudatoolkit = cudaPackages.cudatoolkit`

Side note 2: which actually seems to be broken, now that I look at it

```console
❯ nix shell --impure --expr 'with builtins; with import (getFlake github:NixOS/nixpkgs/master) { config = { allowUnfree = true; cudaSupport = true; }; }; cudaPackages.nsight_systems' --command 'nsys-ui'
Error: Nsight Systems #VERSION_RSPLIT# hasn't been installed with CUDA Toolkit #CUDA_MAJOR#.#CUDA_MINOR#
```

**Comment by mcwitt** [comment]
@samuela Ah, sorry! I had actually been running `nix-shell ./. -p cudatoolkit --command nsys-ui` from a local nixpkgs checkout, and so I assumed I was using master. But it seems like this did not have the intended effect (https://github.com/NixOS/nix/issues/459) and I was instead using the system nixpkgs channel version (which is 897876e).

Using the correct command,

```
NIX_PATH=nixpkgs=. nix-shell -p cudatoolkit --command nsys-ui
```

it works as expected! So it seems as if this has already been fixed in master. Thanks for your response, and apologies for the noise.

**Comment by mcwitt** [comment]
> Side note 2: which actually seems to be broken, now that I look at it
> 
> ```
> ❯ nix shell --impure --expr 'with builtins; with import (getFlake github:NixOS/nixpkgs/master) { config = { allowUnfree = true; cudaSupport = true; }; }; cudaPackages.nsight_systems' --command 'nsys-ui'
> Error: Nsight Systems #VERSION_RSPLIT# hasn't been installed with CUDA Toolkit #CUDA_MAJOR#.#CUDA_MINOR#
> ```

Got it. I did actually try that first, and ran into the same error.

**Comment by SomeoneSerge** [comment]
@mcwitt mind updating the title then? `s/cudatoolkit/cudaPacakges.nsight_systems/`

**Comment by mcwitt** [comment]
I think I've found a separate issue; the `nsys-ui` binary crashes when trying to open a saved profile. I can open a separate issue for this if it makes sense.

Reproduction below:

<details>
Copy the following `flake.nix`

``` nix
{
  inputs.nixpkgs.url = "github:nixos/nixpkgs";

  outputs = { self, nixpkgs }:
    let
      system = "x86_64-linux";

      pkgs = import nixpkgs {
        inherit system;
        config.allowUnfree = true;
      };

      inherit (pkgs) cudaPackages;
    in
    {
      devShells.${system}.default = pkgs.mkShell {
        packages = [
          cudaPackages.cudatoolkit
          self.packages.${system}.saxpy
        ];
      };

      packages.${system} = rec {
        default = saxpy;

        saxpy = pkgs.stdenv.mkDerivation {
          name = "saxpy";
          src = ./saxpy.cu;
          buildInputs = with cudaPackages; [ cuda_nvcc cuda_cudart ];
          dontUnpack = true;
          buildPhase = ''
            nvcc -o saxpy $src
          '';
          installPhase = ''
            mkdir -p $out/bin
            mv saxpy $out/bin/saxpy
          '';
        };
      };
    };
}
```

1. `nix build` to compile an example CUDA program
1. `nix develop` to enter a dev shell with Nsight Systems installed (from `cudatoolkit`, since the redist `cudaPackages.nsight_systems` seems broken)
1. `nsys profile ./result/bin/saxpy` to generate a profile
1. `nsys-ui ./report1.nsys-rep` to attempt to open the profile

Fails with the following dialog (and no useful output to stderr or stdout):
![image](https://github.com/NixOS/nixpkgs/assets/319411/af10770e-eadc-4f7f-9e87-0bc131496cc8)
</details>

**Comment by mcwitt** [comment]
Update: I looked at this a bit and was able to patch the initial issue with this:

https://github.com/mcwitt/nixpkgs/commit/954c43edd05cfaebf43a197b90677633af708609

With this, I'm able to launch the UI when using `cudaPackages.nsight_systems`. However, the UI still crashes when trying to load a profile as described in https://github.com/NixOS/nixpkgs/issues/232458#issuecomment-1555095211. At least the behavior is now the same between `cudatoolkit` and `cudaPackages.nsight_systems`.

**Comment by tehwalris** [comment]
> I think I've found a separate issue; the `nsys-ui` binary crashes when trying to open a saved profile. I can open a separate issue for this if it makes sense.
> 
> Reproduction below:

I get this same crash. Here's my [Minidump file](https://github.com/NixOS/nixpkgs/files/11638328/f50d3807-396a-4e20-88eca7a6-98e01d72.dmp.gz) and a [plain text version](https://github.com/NixOS/nixpkgs/files/11638323/crashdump.txt)
 of it. I don't understand enough to be sure, but it looks like a crash related to the QT web view.

**Comment by graham33** [comment]
@mcwitt @tehwalris please try my fix in #271923 , it worked for and I was able to run your reproduction above successfully.

**Comment by samuela** [comment]
> Unsure how to check exact commits, however.

`github:NixOS/nixpkgs/<commit>` should do the trick

**Comment by tmplt** [comment]
Thanks.

Broken on 44dbbcea48ff4fd6337a8376519ebc4219927229 (master) and 6458f9f1f4f3c7706d3f2ff15ea475df17227d8a (from #271923 that closed this issue).

Same goes for 8ddc8e335530af3d1394fe323838ade3b0a7d2f6.

**Comment by graham33** [comment]
This looks like a different issue to the one I fixed in #271923 . If I have some time in the next few days I'll try to take a look.

**Comment by mcwitt** [comment]
> Thanks.
> 
> Broken on [44dbbce](https://github.com/NixOS/nixpkgs/commit/44dbbcea48ff4fd6337a8376519ebc4219927229) (master) and [6458f9f](https://github.com/NixOS/nixpkgs/commit/6458f9f1f4f3c7706d3f2ff15ea475df17227d8a) (from #271923 that closed this issue).
> 
> Same goes for [8ddc8e3](https://github.com/NixOS/nixpkgs/commit/8ddc8e335530af3d1394fe323838ade3b0a7d2f6).



> This looks like a different issue to the one I fixed in #271923 . If I have some time in the next few days I'll try to take a look.

Sorry, it's possible that I created some confusion by discussing 2 distinct issues in this thread:

1. The path issue in the issue description that @tmplt rediscovered. As far as I know, this has not been fixed, so it seems like this issue should remain open
2. The crash due to Qt libraries mentioned in a comment (https://github.com/NixOS/nixpkgs/issues/232458#issuecomment-1555095211) and fixed by https://github.com/NixOS/nixpkgs/pull/271923. I think merging that PR improperly closed this issue, although (1) hadn't been addressed

**Comment by mcwitt** [comment]
Update: I think I've managed to fix the path issue for `cudaPackages.nsight_systems` with b31e79d3eaadc6f332b97dd652e61b9f77f5e187, but now am stuck on a (different?) Qt-related issue, which occurs when I try to run `nsys-ui`:

```
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

################# ERROR: CrashReporter #################
CrashReporter

Qt initialization failed

Could not find Qt platform plugin
 - Qt plugin search path: /nix/store/9bgz9p97hvg6s5m4z7amcig6bzm9iw4c-nsight_systems-2023.2.3.1004-bin/nsight-systems/2023.2.3/host-linux-x64
 - Qt plugin search path: /nix/store/9bgz9p97hvg6s5m4z7amcig6bzm9iw4c-nsight_systems-2023.2.3.1004-bin/nsight-systems/2023.2.3/host-linux-x64
Application is going to abort
sh: line 1: xmessage: command not found
/nix/store/9bgz9p97hvg6s5m4z7amcig6bzm9iw4c-nsight_systems-2023.2.3.1004-bin/nsight-systems/2023.2.3/host-linux-x64/nsys-ui: line 52: 611912 Aborted                 (core dumped) "$NV_AGORA_PATH/CrashReporter" --hide-stack "NVIDIA Nsight Systems" "NsightSystems" "2023.2.3 (Build 2023.2.3.1004-33186433v0)" "$NV_QUADD_PATH" "$@"
```

@graham33 , since you fixed Qt issues in cudatoolkit, wondering if you have an idea here?

**Comment by graham33** [comment]
> @graham33 , since you fixed Qt issues in cudatoolkit, wondering if you have an idea here?
> 
> It seems like somehow the Qt plugin search path is not set correctly by `wrapQtAppsHook`.

Yes, my fix in #271923 was for the monolithic `cudaPackages.cudatoolkit`, not for the separate `cudaPackages.nsight_systems` redist package with this issue is about, sorry for the confusion! It should be possible to use the same fix for `cudaPackages.nsight_systems`, I'll see if I can give that a try. I can confirm that `nsys-ui` still works for me from `cudaPackages.cudatoolkit` on master, so that's probably a workaround in the interim.

**Comment by SomeoneSerge** [comment]
> `sh: line 1: xmessage: command not found`

```
❯ nix-locate bin/xmessage
xorg.xmessage.out                                16,056 x /nix/store/j4cjh5zg8xn6xg60i5jp5kq4jadx49zv-xmessage-1.0.6/bin/xmessage
```

Has wrapQtAppsHook got anything to do with this?

**Comment by graham33** [comment]
@mcwitt try this commit: https://github.com/graham33/nixpkgs/commit/1ed0f5c959947d2d707d991a4804438ef73eb204- it seemed to fix it for me

**Comment by mcwitt** [comment]
> @mcwitt try this commit: [graham33@1ed0f5c](https://github.com/graham33/nixpkgs/commit/1ed0f5c959947d2d707d991a4804438ef73eb204) - it seemed to fix it for me

Can confirm this fixes the issue for me. Do you think we should PR this?

**Comment by mcwitt** [comment]
@SomeoneSerge 

> Has wrapQtAppsHook got anything to do with this?

Oh hmm, I hadn't even noticed the xmessage error. In any case, this seems to be a red herring, as https://github.com/graham33/nixpkgs/commit/1ed0f5c959947d2d707d991a4804438ef73eb204 seems to fix the issue for me.

**Comment by SomeoneSerge** [comment]
> Oh hmm, I hadn't even noticed the xmessage error. In any case, this seems to be a red herring, as https://github.com/graham33/nixpkgs/commit/1ed0f5c959947d2d707d991a4804438ef73eb204 seems to fix the issue for me.

Has this been merged?

**Comment by graham33** [comment]
> > @mcwitt try this commit: [graham33@1ed0f5c](https://github.com/graham33/nixpkgs/commit/1ed0f5c959947d2d707d991a4804438ef73eb204) - it seemed to fix it for me
> 
> @graham33 Can confirm this fixes the issue for me. Thanks! Do you think we should PR this?

Yes, do you just want to add it to your PR? Your existing commit is needed too.

**Comment by mcwitt** [comment]
> > > @mcwitt try this commit: [graham33@1ed0f5c](https://github.com/graham33/nixpkgs/commit/1ed0f5c959947d2d707d991a4804438ef73eb204) - it seemed to fix it for me
> > 
> > 
> > @graham33 Can confirm this fixes the issue for me. Thanks! Do you think we should PR this?
> 
> Yes, do you just want to add it to your PR? Your existing commit is needed too.

I've cherry-picked https://github.com/graham33/nixpkgs/commit/1ed0f5c959947d2d707d991a4804438ef73eb204 into https://github.com/NixOS/nixpkgs/pull/291471, and verified that this fixes the nsys-ui binary on my end (it starts and basic functionality seems intact).

@tmplt can you check if https://github.com/NixOS/nixpkgs/pull/291471 works for you?

**Comment by tmplt** [comment]
`nsys`:
```
$ nix shell --impure --expr 'with builtins; with import (getFlake github:mcwitt/nixpkgs/fix/nsight_systems) { config = { allowUnfree = true; cudaSupport = true; }; }; cudaPackages.nsight_systems' --command 'nsys' 
```
runs as expected, but `nsys-ui`:
```
$ nix shell --impure --expr 'with builtins; with import (getFlake github:mcwitt/nixpkgs/fix/nsight_systems) { config = { allowUnfree = true; cudaSupport = true; }; }; cudaPackages.nsight_systems' --command 'nsys-ui'
Warning: Failed to get OpenGL version. OpenGL version 2.0 or higher is required.
OpenGL version is too low (0). Falling back to Mesa software rendering.
/nix/store/iqjvhbs6zi51rly82nim489dpxqsslry-nsight_systems-2023.2.3.1004-bin/nsight-systems/2023.2.3/host-linux-x64/nsys-ui: line 52: 216469 Segmentation fault      "$NV_AGORA_PATH/CrashReporter" --hide-stack "NVIDIA Nsight Systems" "NsightSystems" "2023.2.3 (Build 2023.2.3.1004-33186433v0)" "$NV_QUADD_PATH" "$@"
```

I'm in a [WSL2 environment](https://github.com/nix-community/NixOS-WSL), however.

**Comment by SomeoneSerge** [comment]
@tmplt do we remove the vendored `libGL.so` that comes with the archive? That's likely broken as well, we should use `libGL` from nixpkgs

**Comment by mcwitt** [comment]
Superseded by https://github.com/NixOS/nixpkgs/pull/355470

**Comment by mcwitt** [comment]
Accidentally closed (meant to close the related PR)

**Comment by Narsil** [comment]
Hi, I've encountered this too today.

PR https://github.com/NixOS/nixpkgs/pull/355470 fixes `nsys` but not `nsys-ui`. `nsys-ui` still complains about QT.

I've managed to somehow improve the situation slightly:
https://github.com/NixOS/nixpkgs/compare/master...Narsil:nixpkgs:fix/nsight
(Not creating a PR  since it's not functional yet).

Several problems I've encounterd:
- The library `libQtPropertyBrowser.so` cannot be patched by patchelf, therefore I change the rm script from `libQt*` to `libQt6*`.
- The QT plugins are not properly replaced. The  `nsys-ui` binary is now a shell script, and therefore I changed  `dontWrapQtApps=true;` and added `QtWrapApp` on the proper binary.
- This fixes the plugin issues, however when launching there's a missing symbol in `libqssl....` (the exact symbol I forgot but something `isBlackListed` which is defined in `qtbase/...` normally, but I couldn't find any trace of it in nixpkgs. To fix I simply delete that Plugin from the output.

Now the binary properly launches, but still fails because it cannot resolve the OpenGL driver, which is considers to be version `0` even if the string is resolved as `OpenGL ES 3.2 NVIDIA 565.77` (I am running on `nvidiaPackages.beta`).

I'm trying to see if other redist version work better, but I don't think I fully understand how `nsys-ui.bin` actually resolves openGL. I think having version 3.2 should enable it to work.

Note: Running on wayland.

**Comment by Narsil** [comment]
Edit: It actually works with `QT_QPA_PLATFORM=xcb`  (even though I'm on wayland)

**Comment by Narsil** [comment]
I'm trying to also upgrade the version, but the new version has a new issue where Plugins/libQuadDPlugin.so doesn't get patched with some libstc++.so (Cxx11) symbols.

https://github.com/NixOS/nixpkgs/pull/384560

**Comment by Narsil** [comment]
New version works (in the above PR). I fixed nsight_compute in a similar fashion.

For the curious: libQuadDPlugin was fixed by removing all libQt librairies from nsight_compute as original, it's not a slim as it could have, but the missing symbol was `_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE10_M_replaceEmmPKcm@Qt_6` which is a  symbol in libQtWebCore  extremely similar to `_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE10_M_replaceEmmPKcm@@GLIBCXX_3.4.21` which is the real stdlibc++ symbol leading to a lot of confusion for me.
Since it doesn't look like the Qt version from nix is packaging that particular symbol I instead resorted to leaving their custom plugins which are packaged in nsight. This works on later versions  (2024.6.2) because they removed the HARD failure when OpenGL isn't detected as a proper version. I don't think it works on (2023.4.4). I haven't double checked but I remember trying and the app would refuse to launch.

**Comment by prusnak** [comment]
I think this was fixed by https://github.com/NixOS/nixpkgs/pull/393451 which contains the fix by @Narsil 

If not please reopen

**Comment by bigmat18** [comment]
Hello everyone, sorry for the inconvenience but I have encountered a problem similar to @mcwitt regarding the crash of nsys-ui when opening a report file. These are my [minidump](https://drive.google.com/file/d/1ytHopb7bj-DgDUrukjD2H7Ic3MO_g58P/view?usp=sharing) and [txt](https://drive.google.com/file/d/1pWGp76MKYGVToZ6Qo1LQDan4RzLTgeB_/view?usp=sharing) files. 

Additionally, by opening nsys-ui with `strace -o test.txt -f nsys-ui`, I am able to get the following error [report](https://drive.google.com/file/d/18IoOn8LDX5lHon1vUBpj7RP_nMVjgBDR/view?usp=sharing).


I have CUDA version: 12.8. 
Thank you for your help!
